lib/compiler-singlepass/src/machine.rs

```diff
@@ -466,12 +443,51 @@ impl Machine {
             }
         }
 
+        // Save R15 for vmctx use.
+        self.stack_offset.0 += 8;
+        a.emit_mov(
+            Size::S64,
+            Location::GPR(GPR::R15),
+            Location::Memory(GPR::RBP, -(self.stack_offset.0 as i32)),
+        );
+
+        if calling_convention == CallingConvention::WindowsFastcall {
+            for reg in [GPR::RDI, GPR::RSI] {
+                self.stack_offset.0 += 8;
+                a.emit_mov(
+                    Size::S64,
+                    Location::GPR(reg),
+                    Location::Memory(GPR::RBP, -(self.stack_offset.0 as i32)),
+                );
+            }
+        }
+
         // Load vmctx into R15.
         a.emit_mov(
             Size::S64,
             Self::get_param_location(0, calling_convention),
             Location::GPR(GPR::R15),
         );
+    }
+
+    pub(crate) fn init_locals<E: Emitter>(
+        &mut self,
+        a: &mut E,
+        n: u32,
+        n_params: u32,
+        calling_convention: CallingConvention,
+    ) {
+        let locals_size = ((n - n_params) as usize).saturating_sub(Self::LOCAL_REGISTERS.len()) * 8;
```

Comment:
Note to self: look at this and `save_area_offset` handling more closely next time I look at the code.